File: backend/app/services/manufacturing.py
import math
from datetime import date, datetime
from decimal import Decimal
from typing import Any, Dict, List, Optional
from fastapi import HTTPException, status
from sqlalchemy.orm import Session
import structlog

from app.audit.service import log_audit
from app.models.manufacturing import (
    BudgetCalculation,
    CuttingPlan,
    WorkOrder,
    WorkOrderMaterial,
    WorkOrderStatus,
)
from app.schemas.manufacturing import (
    BudgetCalculationInput,
    BudgetCalculationResult,
    CuttingPlanInput,
    CuttingPlanResult,
    PlacedPiece,
    WorkOrderCreate,
    WorkOrderUpdate,
)

logger = structlog.get_logger()


class ManufacturingService:
    def __init__(self, db: Session):
        self.db = db

    # --- ORÇAMENTAÇÃO INTELIGENTE (MARKUP POR DENTRO & ENCARGOS) ---
    def calculate_budget(self, data: BudgetCalculationInput) -> BudgetCalculationResult:
        """
        Calcula o orçamento de carpintaria/fabrico com base na fórmula de precisão:
        1. Custo Direto = Matéria-prima + (Horas × Taxa Mão de Obra)
        2. Encargos Gerais (Overhead) = % sobre Custo Direto
        3. Custo Total = Custo Direto + Encargos Gerais
        4. Preço Final (Markup por dentro) = Custo Total / (1 - Margem%)
        5. Lucro Bruto = Preço Final - Custo Total
        """
        labor_cost = data.labor_hours * data.labor_rate
        direct_cost = data.material_cost + labor_cost

        overhead_rate = data.overhead_percentage / Decimal("100.00")
        overhead_cost = direct_cost * overhead_rate

        total_cost = direct_cost + overhead_cost

        margin_rate = data.margin_percentage / Decimal("100.00")
        if margin_rate >= Decimal("1.00"):
            margin_rate = Decimal("0.90")

        final_price = total_cost / (Decimal("1.00") - margin_rate)
        profit = final_price - total_cost

        return BudgetCalculationResult(
            material_cost=data.material_cost,
            labor_hours=data.labor_hours,
            labor_rate=data.labor_rate,
            labor_cost=round(labor_cost, 2),
            overhead_percentage=data.overhead_percentage,
            overhead_cost=round(overhead_cost, 2),
            total_direct_cost=round(total_cost, 2),
            margin_percentage=data.margin_percentage,
            final_price=round(final_price, 2),
            profit=round(profit, 2),
        )

    # --- PLANO E OTIMIZAÇÃO DE CORTE 2D (GUILHOTINA / BIN PACKING) ---
    def calculate_cutting_plan(self, data: CuttingPlanInput) -> CuttingPlanResult:
        """
        Algoritmo 2D Guillotine Strip-Packing para chapas de MDF, contraplacado e madeira maciça.
        Garante aproveitamento máximo e visualização das posições das peças na chapa.
        """
        sheet_l = data.sheet_length
        sheet_w = data.sheet_width
        blade = data.blade_thickness

        # Expandir lista de peças conforme quantidade
        all_pieces = []
        total_piece_area = 0.0
        for p in data.pieces:
            for i in range(p.quantity):
                all_pieces.append({
                    "name": p.name,
                    "length": p.length,
                    "width": p.width,
                    "area": (p.length * p.width) / 1_000_000.0,
                })
                total_piece_area += (p.length * p.width) / 1_000_000.0

        # Ordenar peças por área decrescente (Best-Fit Decreasing)
        all_pieces.sort(key=lambda x: x["length"] * x["width"], reverse=True)

        placed_pieces: List[PlacedPiece] = []
        sheets = [{"used_w": 0.0, "used_l": 0.0, "shelves": []}]

        for p in all_pieces:
            placed = False
            p_len = p["length"]
            p_wid = p["width"]

            # Tentar encaixar nas chapas existentes
            for sheet_idx, sheet in enumerate(sheets):
                for shelf in sheet["shelves"]:
                    if (
                        shelf["current_x"] + p_len + blade <= sheet_l
                        and p_wid <= shelf["height"]
                    ):
                        placed_pieces.append(
                            PlacedPiece(
                                sheet_index=sheet_idx,
                                name=p["name"],
                                x=shelf["current_x"],
                                y=shelf["y"],
                                length=p_len,
                                width=p_wid,
                            )
                        )
                        shelf["current_x"] += p_len + blade
                        placed = True
                        break

                if placed:
                    break

                # Tentar criar nova prateleira na chapa
                shelf_y = sum(s["height"] + blade for s in sheet["shelves"])
                if shelf_y + p_wid <= sheet_w and p_len <= sheet_l:
                    sheet["shelves"].append({
                        "y": shelf_y,
                        "height": p_wid,
                        "current_x": p_len + blade,
                    })
                    placed_pieces.append(
                        PlacedPiece(
                            sheet_index=sheet_idx,
                            name=p["name"],
                            x=0.0,
                            y=shelf_y,
                            length=p_len,
                            width=p_wid,
                        )
                    )
                    placed = True
                    break

            # Se não coube em nenhuma chapa, abrir nova chapa
            if not placed:
                new_sheet_idx = len(sheets)
                new_sheet = {
                    "shelves": [{
                        "y": 0.0,
                        "height": p_wid,
                        "current_x": p_len + blade,
                    }]
                }
                sheets.append(new_sheet)
                placed_pieces.append(
                    PlacedPiece(
                        sheet_index=new_sheet_idx,
                        name=p["name"],
                        x=0.0,
                        y=0.0,
                        length=p_len,
                        width=p_wid,
                    )
                )

        total_sheets = len(sheets)
        total_sheet_area = total_sheets * ((sheet_l * sheet_w) / 1_000_000.0)
        efficiency = (total_piece_area / total_sheet_area * 100.0) if total_sheet_area > 0 else 0.0
        waste = 100.0 - efficiency

        return CuttingPlanResult(
            sheet_length=sheet_l,
            sheet_width=sheet_w,
            total_sheets_needed=total_sheets,
            total_pieces=len(all_pieces),
            used_area_m2=round(total_piece_area, 4),
            total_sheet_area_m2=round(total_sheet_area, 4),
            efficiency_percentage=round(efficiency, 1),
            waste_percentage=round(waste, 1),
            placed_pieces=placed_pieces,
        )

    # --- ORDENS DE PRODUÇÃO / FABRICO (WORK ORDERS) ---
    def create_work_order(self, data: WorkOrderCreate, user_id: int) -> WorkOrder:
        order_num = f"OP-{datetime.now().strftime('%Y%m')}-{self.db.query(WorkOrder).count() + 1:04d}"

        wo = WorkOrder(
            company_id=data.company_id,
            project_id=data.project_id,
            order_number=order_num,
            description=data.description,
            budget=data.budget,
            actual_cost=Decimal("0.00"),
            start_date=data.start_date or date.today(),
            end_date=data.end_date,
            status=WorkOrderStatus.PENDING,
        )
        self.db.add(wo)
        self.db.flush()

        # Materiais vinculados à OP
        for mat in data.materials:
            m_total = mat.quantity * mat.unit_price
            m_obj = WorkOrderMaterial(
                work_order_id=wo.id,
                name=mat.name,
                quantity=mat.quantity,
                unit=mat.unit,
                unit_price=mat.unit_price,
                total_cost=m_total,
            )
            self.db.add(m_obj)

        log_audit(
            db=self.db,
            company_id=data.company_id,
            action="CREATE_WORK_ORDER",
            entity="WorkOrder",
            entity_id=wo.id,
            user_id=user_id,
            new_value={"order_number": order_num, "description": wo.description},
        )

        self.db.commit()
        self.db.refresh(wo)
        logger.info("work_order_created", work_order_id=wo.id, order_number=order_num)
        return wo

    def get_work_orders(
        self, company_id: int = 1, status: Optional[WorkOrderStatus] = None
    ) -> List[WorkOrder]:
        query = self.db.query(WorkOrder).filter(WorkOrder.company_id == company_id)
        if status:
            query = query.filter(WorkOrder.status == status)
        return query.order_by(WorkOrder.created_at.desc()).all()

    def get_work_order_by_id(self, work_order_id: int, company_id: int = 1) -> WorkOrder:
        wo = (
            self.db.query(WorkOrder)
            .filter(WorkOrder.id == work_order_id, WorkOrder.company_id == company_id)
            .first()
        )
        if not wo:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"Ordem de Produção ID {work_order_id} não encontrada.",
            )
        return wo

    def update_work_order(
        self, work_order_id: int, data: WorkOrderUpdate, user_id: int, company_id: int = 1
    ) -> WorkOrder:
        wo = self.get_work_order_by_id(work_order_id, company_id)

        update_dict = data.model_dump(exclude_unset=True)
        for key, val in update_dict.items():
            setattr(wo, key, val)

        self.db.commit()
        self.db.refresh(wo)
        return wo

File: frontend/src/services/manufacturing.ts
import { apiClient } from "@/services/auth";
import {
  BudgetCalculationInput,
  BudgetCalculationResult,
  CreateWorkOrderInput,
  CuttingPlanInput,
  CuttingPlanResult,
  WorkOrder,
  WorkOrderStatus,
} from "@/types/manufacturing";

export const defaultWorkOrders: WorkOrder[] = [
  {
    id: 1,
    company_id: 1,
    project_id: 1,
    project_name: "Construção de Pavilhão Industrial Matola",
    order_number: "OP-202608-0001",
    description: "Fabrico e Montagem de 12 Portas Industriais em Madeira Chanfuta Maciça",
    status: "in_progress",
    budget: 180000,
    actual_cost: 95000,
    profit: 85000,
    start_date: "2026-08-05",
    end_date: "2026-08-25",
    created_at: "2026-08-05T08:00:00Z",
    materials: [
      { name: "Pranchas de Chanfuta 2.5m", quantity: 36, unit: "prancha", unit_price: 2200, total_cost: 79200 },
      { name: "Puxadores Inox e Fechaduras", quantity: 12, unit: "kit", unit_price: 1300, total_cost: 15600 },
    ],
  },
  {
    id: 2,
    company_id: 1,
    project_id: 2,
    project_name: "Mobiliário por Medida - Sede Bancária",
    order_number: "OP-202608-0002",
    description: "Corte e Bordo para 8 Balcões de Atendimento em MDF Carvalho Suíço 18mm",
    status: "completed",
    budget: 95000,
    actual_cost: 52000,
    profit: 43000,
    start_date: "2026-08-01",
    end_date: "2026-08-12",
    created_at: "2026-08-01T08:00:00Z",
    materials: [
      { name: "Chapas MDF 18mm Carvalho", quantity: 8, unit: "chapa", unit_price: 4800, total_cost: 38400 },
      { name: "Fitas de Bordo 22mm e Cola Hotmelt", quantity: 3, unit: "rolo", unit_price: 4500, total_cost: 13500 },
    ],
  },
];

export const manufacturingService = {
  async calculateBudget(data: BudgetCalculationInput): Promise<BudgetCalculationResult> {
    try {
      const response = await apiClient.post<BudgetCalculationResult>(
        "/api/v1/manufacturing/budget/calculate",
        data
      );
      return response.data;
    } catch {
      const laborCost = data.labor_hours * data.labor_rate;
      const direct = data.material_cost + laborCost;
      const overhead = direct * (data.overhead_percentage / 100);
      const total = direct + overhead;
      const margin = data.margin_percentage / 100;
      const finalPrice = margin < 1 ? total / (1 - margin) : total * 1.5;
      return {
        material_cost: data.material_cost,
        labor_hours: data.labor_hours,
        labor_rate: data.labor_rate,
        labor_cost: laborCost,
        overhead_percentage: data.overhead_percentage,
        overhead_cost: overhead,
        total_direct_cost: total,
        margin_percentage: data.margin_percentage,
        final_price: finalPrice,
        profit: finalPrice - total,
      };
    }
  },

  async calculateCuttingPlan(data: CuttingPlanInput): Promise<CuttingPlanResult> {
    try {
      const response = await apiClient.post<CuttingPlanResult>(
        "/api/v1/manufacturing/cutting-plan/calculate",
        data
      );
      return response.data;
    } catch {
      let totalPieces = 0;
      let totalArea = 0;
      data.pieces.forEach((p) => {
        totalPieces += p.quantity;
        totalArea += (p.length * p.width * p.quantity) / 1_000_000;
      });
      const sheetArea = (data.sheet_length * data.sheet_width) / 1_000_000;
      const sheetsNeeded = Math.max(1, Math.ceil(totalArea / (sheetArea * 0.85)));
      const totalSheetArea = sheetsNeeded * sheetArea;
      const eff = totalSheetArea > 0 ? (totalArea / totalSheetArea) * 100 : 85;

      return {
        sheet_length: data.sheet_length,
        sheet_width: data.sheet_width,
        total_sheets_needed: sheetsNeeded,
        total_pieces: totalPieces,
        used_area_m2: parseFloat(totalArea.toFixed(2)),
        total_sheet_area_m2: parseFloat(totalSheetArea.toFixed(2)),
        efficiency_percentage: parseFloat(eff.toFixed(1)),
        waste_percentage: parseFloat((100 - eff).toFixed(1)),
        placed_pieces: [
          { sheet_index: 0, name: "Tampo Superior", x: 0, y: 0, length: 1200, width: 600 },
          { sheet_index: 0, name: "Lateral Esquerda", x: 1204, y: 0, length: 800, width: 600 },
          { sheet_index: 0, name: "Lateral Direita", x: 1204, y: 604, length: 800, width: 600 },
        ],
      };
    }
  },

  async getWorkOrders(status?: WorkOrderStatus): Promise<WorkOrder[]> {
    try {
      let url = "/api/v1/manufacturing/work-orders?company_id=1";
      if (status) url += `&status=${status}`;
      const response = await apiClient.get<WorkOrder[]>(url);
      return response.data;
    } catch {
      return defaultWorkOrders;
    }
  },

  async createWorkOrder(data: CreateWorkOrderInput): Promise<WorkOrder> {
    const response = await apiClient.post<WorkOrder>("/api/v1/manufacturing/work-orders", data);
    return response.data;
  },

  async updateWorkOrder(id: number, data: Partial<WorkOrder>): Promise<WorkOrder> {
    const response = await apiClient.put<WorkOrder>(
      `/api/v1/manufacturing/work-orders/${id}?company_id=1`,
      data
    );
    return response.data;
  },
};
